### ISSUE TITLE
Migration failing with error relation does not exist

### ISSUE THREAD

**Issue by privateOmega** [knex/knex#3115]
# Environment

Knex version: 0.16.3
Database + version: PostgreSQL  10.5
OS: Windows

# Bug

1. Migration failing with error message `alter table "meraki"."role_permissions" add constraint "role_permissions_role_id_foreign" foreign key ("role_id") references "roles" ("id") - relation "roles" does not exist`
2. https://gist.github.com/privateOmega/18bff0b56e70d4260f8c0d0897134ca8

Discussion on SO can be found [here](https://stackoverflow.com/questions/55277791/issue-with-knex-migration)

**Comment by elhigu** [comment]
So according to stackoverflow problem is that withSchema value is not appended to `.inTable()` declaration?

**Comment by privateOmega** [comment]
@elhigu Yup exactly. Just like Fazal answered, we need to prepend schema table with table name in inTable.

**Comment by elhigu** [comment]
Thanks! I think that the correct functionality would be to default schema to table’s schema and to document that behaviour where schema is allowed to be passed to inTable.

**Comment by MarcusCemes** [comment]
I've spent the last hour thinking it's a problem with the table not yet existing because it's in the same transaction (seems dumb now, I know). This definitely needs to be documented.

**Comment by kibertoad** [comment]
No, it needs to be fixed :-D

**Comment by privateOmega** [comment]
This issue totally slipped my mind, will put in a PR ASAP either to document or add default behaviour to fall back to specified schema.

**Comment by kibertoad** [comment]
I would much prefer not to document it, because it is wrong behaviour. withSchema should be applied to everything.

**Comment by privateOmega** [comment]
@kibertoad that's very true.

**Comment by nmaxcom** [comment]
I am having the same error showing up with migrate:latest. How can I get passed it? I've tried with `.schema.withSchema('public')` and also `table.foreign('author').references('id').inTable('public.users');` but I still get pretty much the same:

> migration failed with error: alter table "public"."newsArticles" add constraint "newsarticles_author_foreign" foreign key ("author") references "public"."users" ("id") - relation "public.users" does not exist

> error: relation "public.users" does not exist

How to get passed this?!

**Comment by elgatoflaco** [comment]
I recently had a similar problem and fixed it by editing the model manually.
In my case spaces had a many to many relationship with spaceCategories and did not recognize the relationship, so I edited the model manually, adding
`"collectionName": "spaces_space_categories__space_categories_space"`

<img width="675" alt="Captura de pantalla 2021-01-19 a las 10 12 06" src="https://user-images.githubusercontent.com/10778910/105015084-752a1b80-5a41-11eb-9561-7892e2ff16ce.png">

**Comment by marceliwac** [comment]
I can confirm that I am also running into a similar issue (possibly related), where running two consecutive migrations using `migrate:latest` fails, as the second one throws an error. The first one is responsible for building schema for all tables, which works fine, but the latter which attempts to insert into those tables, fails with "relation xxx does not exist".

**Comment by talesmgodois** [comment]
What if I want just to use the table name? I am using a database that I didnt create the migration using knex

**Comment by mercmobily** [comment]
I looked at it. This seems to be a confirmed bug in how schema is applied for foreign keys on Postgres.

Previous maintainers stated clearly that this should be fixed, not merely documented. Specifically, `.withSchema()` on a table builder should apply to foreign key targets by default. elhigu suggested the correct behaviour is to default the referenced table’s schema to the table’s own schema when none is specified; kibertoad confirmed “it needs to be fixed”.

What happens today: `table.withSchema('meraki')` sets the schema for the table being created/altered, but `table.foreign(...).references('roles.id')` does not inherit that schema for the referenced table. Unless you explicitly qualify the target (e.g. `.references('meraki.roles.id')` or `.inTable('meraki.roles')`), Postgres sees `roles` without a schema and errors with “relation roles does not exist” when your `search_path` does not include it.

Evidence in code:
- lib/schema/tablebuilder.js:90–117 collects `foreignData.inTable` from `references(tableColumn)` and `on(tableName)` as a bare string. There is no logic to inject the current builder’s schema (`this._schemaName`) when `inTable` is unqualified.
- lib/schema/tablecompiler.js: foreign(foreignData) compiles `inTable` via `this.formatter.wrap(foreignData.inTable)`, which wraps whatever string was provided, but does not prefix the builder’s schema. In contrast, the table name for the DDL itself comes from `this.tableName()` which does consider `this.schemaNameRaw`.
- Postgres compiler uses `this.tableName()` for the left‑hand side and relies on the already‑wrapped `inTable` for the right‑hand side, so the mismatch persists (lib/dialects/postgres/schema/pg-tablecompiler.js: general FK add paths).

Reproduction outline (matches the issue):
- Create schema `meraki`.
- `knex.schema.withSchema('meraki').createTable('role_permissions', (t) => t.integer('role_id').references('id').inTable('roles'));`
- Unless `roles` is on `search_path` or you fully qualify it, FK DDL becomes `... references "roles" ("id") ...` and fails with relation does not exist.

Workarounds until fixed:
- Fully qualify the reference: `.references('meraki.roles.id')` or `.references('id').inTable('meraki.roles')`.
- Alternatively, ensure `search_path` contains the referenced schema for the migration session (less explicit, more fragile).

Proposed fix (backwards‑compatible and aligned with maintainer guidance):
- When building `foreignData.inTable`, if the value is unqualified (no dot) and `tableBuilder._schemaName` is set, qualify it as `${schema}.${table}` before compiling. This preserves existing behaviour for already qualified inputs and for no‑schema cases.
- Apply the same rule to `references('table.column')` where the table part is present but unqualified; default the schema there as well.
- Add tests in Postgres suite covering `.withSchema('x')` on a table and unqualified FK references resolving to `x.table`.

**Comment by thecoolestcoderrushil** [comment]
migrations that pass locally but fail at deploy with relation does not exist are frustrating, since the schema setup differs between environments. rehearsing the migration on a production-shaped twin first shows these failures before they block a deploy. antifailure.dev (https://antifailure.dev) builds disposable postgres twins for this kind of migration rehearsal and is open source and free to try, with demos available for teams that want one.
